### PR TITLE
Feat/signalen classification integration in django admin

### PR DESCRIPTION
**Pull request by justiandevs**
## Description

Signalen classification integration into django admin.

### CODE REVIEW

File: app/signals/apps/classification/admin/admins.py [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+
+class TrainingSetAdmin(admin.ModelAdmin):
+    list_display = ('name', 'file', )
+    actions = ["run_training_with_dataset"]
+
+    @admin.action(description="Train model met geselecteerde dataset")
+    def run_training_with_dataset(self, request, queryset):
+        """
+        TODO: run actual training with dataset
+        """
```
Comment (justiandevs) [inline app/signals/apps/classification/admin/admins.py:12]: Feedback van Yuri: Sanity check, checken of alle kolommen geldig zijn. Dataset verifieren. 

Het zou mooi zijn als je feedback op basis van de dataset kan geven.

File: app/signals/apps/classification/models/training_set.py [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+from signals.apps.services.domain.checker_factories import ContentCheckerFactory
+from signals.apps.services.domain.mimetypes import MimeTypeFromContentResolverFactory, \
+    MimeTypeFromFilenameResolverFactory
+from signals.apps.services.validator.file import MimeTypeAllowedValidator, MimeTypeIntegrityValidator, \
+    ContentIntegrityValidator, FileSizeValidator
+
+
+class TrainingSet(models.Model):
+    """
+    This model represents a training set consisting of a single XLSX file, with a "Main", "Middle", "Sub" and "Text" column
```
Comment (bartjkdp) [inline app/signals/apps/classification/models/training_set.py:13]: Ik zou naar een structuur toegaan die overeen komt met het model in Signalen:

- main
- sub
- text

File: app/signals/apps/classification/admin/admins.py [hunk 3]
```diff
@@ -0,0 +1,40 @@
+from django.contrib import admin, messages
+
+
+class TrainingSetAdmin(admin.ModelAdmin):
+    list_display = ('name', 'file', )
+    actions = ["run_training_with_dataset"]
+
+    @admin.action(description="Train model met geselecteerde dataset")
```
Comment (bartjkdp) [inline app/signals/apps/classification/admin/admins.py:8]: Misschien zou je wel alvast een lege Celery task kunnen triggeren, zodat je ook een gevoel krijgt bij hoe async gaat werken.

File: app/signals/apps/classification/admin/admins.py (hunk 1 above)
Comment (bartjkdp) [reply app/signals/apps/classification/admin/admins.py:12]: Lijkt me goed om deze sanity checks synchroon uit te voeren, zodat de functioneel beheerder direct feedback terugkrijgt in Django admin.

Als het bestand goed geformatteerd is, dan kun je een asynchrone job triggeren om het model daadwerkelijk te trainen.

File: app/signals/apps/classification/models/classifier.py [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+    This model represents a classification model consisting of a reference to the "Middle" model and a reference to the
+
+    "Middle, Sub" model
+    """
+    created_at = models.DateTimeField(editable=False, auto_now_add=True)
+
+    name = models.CharField(max_length=255, null=False, blank=False)
+    precision = models.FloatField(null=False, blank=False, default=0)
+    recall = models.FloatField(null=False, blank=False, default=0)
+    accuracy = models.FloatField(null=False, blank=False, default=0)
```
Comment (bartjkdp) [inline app/signals/apps/classification/models/classifier.py:15]: Kan het zo zijn dat je wel al een Classifier hebt, maar dat deze nog niet getraind is (het trainen loopt op dat moment nog asynchroon op de achtergrond)? In dat geval denk ik dat het goed is om het mogelijk te maken dat precision, recall, accuracy en de models ook leeg te laten.

File: app/signals/apps/classification/models/classifier.py [hunk 5]
```diff
@@ -0,0 +1,29 @@
...
+
+    middle_model = models.FileField(
```
Comment (bartjkdp) [inline app/signals/apps/classification/models/classifier.py:17]:
```suggestion
    main_model = models.FileField(
```

File: app/signals/apps/classification/models/classifier.py [hunk 6]
```diff
@@ -0,0 +1,29 @@
...
+        upload_to='classification_models/middle/%Y/%m/%d/',
+        null=False,
+        blank=False,
+        max_length=255,
+    )
+
+    middle_sub_model = models.FileField(
```
Comment (bartjkdp) [inline app/signals/apps/classification/models/classifier.py:24]:
```suggestion
    sub_model = models.FileField(
```

File: app/requirements/requirements.txt [hunk 7]
```diff
@@ -113,32 +113,32 @@ django==4.2.15
...
+    # via -r requirements/requirements.in
 django-storages==1.14.4
-    # via -r requirements.in
+    # via -r requirements/requirements.in
 django-timezone-field==7.0
     # via django-celery-beat
 djangorestframework==3.15.2
     # via
-    #   -r requirements.in
+    #   -r requirements/requirements.in
```
Comment (bartjkdp) [inline app/requirements/requirements.txt:141]: Als je het pip commando draait vanuit de requirements map, dan worden niet al deze changes gedaan. Dat is misschien overzichtelijker.

File: app/signals/apps/classification/tasks.py [hunk 8]
```diff
@@ -0,0 +1,6 @@
+from signals.celery import app
+
+
+@app.task
+def run_training_task():
```
Comment (bartjkdp) [inline app/signals/apps/classification/tasks.py:5]:
```suggestion
def train_classifier(training_set_ids):
```

File: app/signals/apps/classification/tasks.py [hunk 9]
```diff
@@ -0,0 +1,6 @@
...
+    print("running a training task")
```
Comment (bartjkdp) [inline app/signals/apps/classification/tasks.py:6]:
```suggestion
    print("will train a classifier using training sets:", training_set_ids)
```

File: app/signals/apps/classification/admin/admins.py [hunk 10]
```diff
@@ -0,0 +1,76 @@
+from django.contrib import admin, messages
+from signals.apps.classification.tasks import run_training_task
```
Comment (bartjkdp) [inline app/signals/apps/classification/admin/admins.py:2]:
```suggestion
from signals.apps.classification.tasks import train_classifier
```

File: app/signals/apps/classification/admin/admins.py [hunk 11]
```diff
@@ -0,0 +1,76 @@
...
+            if not any(data_rows):
+                self.message_user(
+                    request,
+                    f"The training set { training_set.name } does not contain any data rows.",
+                    messages.ERROR
+                    )
+                return
+
+            # TODO: run actual training task
+            run_training_task.delay()
```
Comment (bartjkdp) [inline app/signals/apps/classification/admin/admins.py:47]:
```suggestion
            train_classifier.delay()
```

**Review by bartjkdp — APPROVED** [review]
Looking good 👍 some minor suggestions.

File: app/signals/apps/classification/tasks.py [hunk 12]
```diff
@@ -1,6 +1,116 @@
...
+    model = grid_search
+
+    # step 5: evaluation
+    test_predict = model.predict(test_texts)
+    precision = str(round(precision_score(test_labels, test_predict, average='macro', zero_division=0), 2))
+    recall = str(round(recall_score(test_labels, test_predict, average='macro'), 2))
+    accuracy = str(round(accuracy_score(test_labels, test_predict), 2))
+
+    # step 6: save model
+    save_dir = datetime.now().strftime('classification_models/main/%Y/%m/%d/')
```
Comment (bartjkdp) [inline app/signals/apps/classification/tasks.py:99]: Je implementeert hier opnieuw logica die eigenlijk al in Django zit. Misschien is dit simpeler:

```
from django.core.files import File

pickled_model = pickle.dumps(model, pickle.HIGHEST_PROTOCOL)

Classifier.objects.create(
    main_model=File(pickled_model)
    ...
```

Comment (justiandevs) [reply app/signals/apps/classification/tasks.py]: Oke, ik heb dat toegepast :)

File: app/signals/apps/classification/train.py [hunk 13]
```diff
@@ -0,0 +1,127 @@
...
+            self.df = pd.read_csv(self.training_set.file, sep=None, engine='python')
+        elif extension == '.xlsx':
+            self.df = pd.read_excel(self.training_set.file)
+        else:
+            raise Exception('Could not read input file. Extension should be .csv or .xlsx')
+
+    def preprocess_file(self):
+        self.df = self.df.dropna(axis=0)
+        self.df["_main_label"] = self.df["Main"]
+        self.df["_sub_label"] = self.df["Main"] + "|" + self.df["Sub"]
```
Comment (bartjkdp) [inline app/signals/apps/classification/train.py:48]: Misschien mooi om template (f) strings te gebruiken hier?

File: app/signals/apps/api/views/ml_tool_proxy_v2.py [hunk 14]
```diff
@@ -0,0 +1,28 @@
...
+
+
+@extend_schema(exclude=True)
+class LegacyMlPredictCategoryViewV2(APIView):
+    def post(self, request, *args, **kwargs):
+        try:
+            classifier = Classifier.objects.get(is_active=True)
+
+            main_model = pickle.load(classifier.main_model)
+            sub_model = pickle.load(classifier.sub_model)
```
Comment (bartjkdp) [inline app/signals/apps/api/views/ml_tool_proxy_v2.py:18]: Hier zouden we denk ik wel een LRU (in-memory) cache mogen gebruiken. Dat zal de performance van het endpoint enorm verhogen. Daar kunnen we samen even naar kijken.

File: app/signals/apps/classification/models/training_set.py [hunk 15]
```diff
@@ -0,0 +1,39 @@
...
+
+class TrainingSet(models.Model):
+    """
+    This model represents a training set consisting of a single XLSX file, with a "Main", "Sub" and "Text" column
+    """
+    created_at = models.DateTimeField(editable=False, auto_now_add=True)
+
+    name = models.CharField(max_length=255, null=False, blank=False)
+
+    file = models.FileField(
```
Comment (bartjkdp) [inline app/signals/apps/classification/models/training_set.py:19]: Hier moeten we nog even de DWH storage gebruiken, in plaats van de media storage. De media storage is namelijk publiek bereikbaar. Dat is niet zo wenselijk voor het ML model.

File: app/signals/apps/classification/models/classifier.py [hunk 16]
```diff
@@ -0,0 +1,30 @@
...
+    """
+    created_at = models.DateTimeField(editable=False, auto_now_add=True)
+
+    name = models.CharField(max_length=255, null=False, blank=False)
+    precision = models.FloatField(null=True, blank=True, default=0)
+    recall = models.FloatField(null=True, blank=True, default=0)
+    accuracy = models.FloatField(null=True, blank=True, default=0)
+    is_active = models.BooleanField(default=False)
+
+    main_model = models.FileField(
```
Comment (bartjkdp) [inline app/signals/apps/classification/models/classifier.py:18]: DWH media storage

File: app/signals/apps/classification/models/classifier.py [hunk 17]
```diff
@@ -0,0 +1,30 @@
...
+        upload_to='classification_models/main/%Y/%m/%d/',
+        null=True,
+        blank=True,
+        max_length=255,
+    )
+
+    sub_model = models.FileField(
```
Comment (bartjkdp) [inline app/signals/apps/classification/models/classifier.py:25]: DWH storage

File: app/signals/apps/api/views/ml_tool_proxy.py [hunk 18]
```diff
@@ -8,6 +8,12 @@
 
 from signals.apps.api.ml_tool.client import MLToolClient
 from signals.apps.signals.models import Category
+import pickle
```
Comment (bartjkdp) [inline app/signals/apps/api/views/ml_tool_proxy.py:11]: Misschien is het mooi (om conflicten te voorkomen) dat we een nieuwe view onderbrengen in signals/apps/classification/views.py.

**Comment by bartjkdp** [comment]
I just squashed merged your work into our release branch release/2.37.10. Thanks 👍  🚀
